### ISSUE TITLE
Feature Request: Localization/Internationalization

### ISSUE THREAD

**Issue by normandc** [alexrj/Slic3r#1378]
Slic3r has been around for a long time now and still has no language support. This would be a highly useful feature for people who struggle with English. Please consider that there are many RepRap communities working in other languages than English.

linvswin has opened a pull request two months ago and didn't even get an acknowledgment. https://github.com/alexrj/Slic3r/pull/1214

I'll be more than willing to provide a French translation once localization is supported by Slic3r.

Please take this into consideration.

**Comment by lordofhyphens** [comment]
Yeah. -.-; I've asked for an update in #1214 (trying to get things going), but nobody's noticed yet. This would be an ongoing thing as well; given any UI changes would prompt more translation.

**Comment by Patola** [comment]
Following this issue for when it is active. As soon as i18n support is added I will add brazilian portuguese translations.

**Comment by lordofhyphens** [comment]
I've been thinking on and off about this and how to solve it in an extensible solution across the 2 languages in use.

One thing that has occurred to me is moving all of the title and description strings to XML (or some other resource file), using the option name as a key.

XML appeals to me as it it should make translation easy and decouples some of the view from the library.

**Comment by lordofhyphens** [comment]
The other thing to do would be to respect the system locale for picking languages (again, defaulting to English if corresponding file doesn't exist).

**Comment by lordofhyphens** [comment]
Since we're using wxWidgets: http://docs.wxwidgets.org/trunk/overview_xrc.html

**Comment by Patola** [comment]
Don't you think simply using gettext would be better? It has a very popular standardized format with lots of specific editors for internationalization. It supports both C++ and perl, the two programming languages (besides tons of others). With XML you would have to deal with the algorithm yourself, with gettext the common problems have already been solved. Also, I think that volunteers to Slic3r that would be willing to translate the strings would be more motivated by gettext .po files than xml.

**Comment by xoan** [comment]
http://docs.wxwidgets.org/3.1/overview_i18n.html

**Comment by Patola** [comment]
@xoan you beat me by one minute. lol

**Comment by gege2b** [comment]
I too think that gettext is the way to go for the translation
but whatever is choosen, I'm up to do french language

**Comment by forReason** [comment]
Im split about translation.
a) yes it is good for people not speaking english
b) I personally HATE using programs in my native Language. The reason being is: I search for an option in photoshop. Find a lot of support online BUT the options there have a different name so I search and search and search.

One premium example would be League of legends where I call my Cousin "Skill Ragnarock" and he will never find it since its called Götterdämmerung in German.

What I think would be a solution is to translate all descriptions and minor things like place on plater but not to mess up much with often used titles such as "Nozzle", "Infill", "Extruder" and so on. For many Things like "Infill" I would not even have an apropriate translation in german.

Nontheless I'm willing to provide german translations

**Comment by Patola** [comment]
@forReason same here. I do translations but I never use them. The reason I need to have the software translated is because I am a 3D printing teacher and most of my students do not have good english. All my bug reports and forum discussions are done with the english version though, and I highlight that fact when teaching my students.

**Comment by lordofhyphens** [comment]
I've produced a basic english .po file for people to copy and use in the i18n branch, it contains just the setting label names and tooltip text from libslic3r at this point. 

Currently there is no support in the GUI for using the translation, but the plan is to employ it when that part of the UI is ported to xsgui. 

If someone wishes to work on a translation, please (PLEASE) issue pull requests against that branch only and it must only touch files in the i18n folder.

**Comment by lordofhyphens** [comment]
branch https://github.com/alexrj/Slic3r/tree/i18n is current place to submit PRs against to add translations for libslic3r strings. As more of the UI is ported to C++, more msgids will get added.

**Comment by lordofhyphens** [comment]
@diegoolavarria

**Comment by lordofhyphens** [comment]
From my experimenting, wxperl's Wx::GetTranslation is broken when using from Wx::Locale. Which is unfortunate.
AFAIK though it should work just fine with xsgui. 

I will be defining a macro in libslic3r.h to mark strings as candidates for translation so we can use poedit's source scanning mechanism to identify them.

**Comment by lordofhyphens** [comment]
For people looking to make translations, I currently recommend poedit (https://poedit.net/).

**Comment by lordofhyphens** [comment]
@forReason @xoan @Patola How do you feel about organizing the translations as a separate repository linked to Slic3r via submodules?

**Comment by Patola** [comment]
Seems complicated. I will have to study git/github to see how this would be
done. Do you have an example of project done this way?

**Comment by lordofhyphens** [comment]
@Patola  Submodules are actually pretty easy to manage. From your POV, you would have a repository that you would commit to. 

People who want to run the latest and greatest would be able to cd into the submodule and update it. 
So the process to get it in the default checkout would be as follows: 

Make changes.
Commit to the translation repo. 

Every so often, @alexrj or I would update the pointer to the current revision in the translation repo. It's possible to write a script that fires when someone commits to the translation repo to update the main repo as well.

The build servers, however, can easily be scripted to grab the new translations immediately.

**Comment by lordofhyphens** [comment]
https://github.com/blog/2104-working-with-submodules

In our use case, the submodule a component of Slic3r and is dependent on its interface.

**Comment by Patola** [comment]
@lordofhyphens thanks for the amazing reference. I will study it. That seems quite a good strategy indeed. But then again, currently one should fork the strings at the i18n branch to do the translations or try and create a submodule?

**Comment by lordofhyphens** [comment]
Until I decide to do it differently, we'll stick with pull requests. If we do a submodule, we might move alexrj/Slic3r to an organizational setup (similar to how Marlin or Smoothieware does it) and do everything at once. I know how to convert to a submodule, so that's not a problem either. 

The master list of strings will increase as more of the GUI is ported from wxperl. 

TL;DR: Fork slic3r's i18n branch and submit pull requests against it.

**Comment by lordofhyphens** [comment]
@powertide see my previous post to this one.

**Comment by powertide** [comment]
I am gonna help to add Chinese translations.
And to do that, I'm going to use Poedit to do translation work, make my ".Po" file  and then commit to i18n branch. Am I gettin it right? @lordofhyphens

**Comment by lordofhyphens** [comment]
@powertide Correct.

**Comment by NateTG** [comment]
Is there any easy way to find the text in context?

**Comment by lordofhyphens** [comment]
Look at PrintConfig.cpp for now.

**Comment by NateTG** [comment]
I was trying to understand what one of the features does.  It looks like the "solid-infill-every-layers" option doesn't actually do anything. Is that correct?

**Comment by SCiunczyk** [comment]
Initially I agree with @forReason, but ... recently I was build small delta printer for my godson. This young guy is only 12 year old and his English is still not so perfect. With polish interface he would focus more on 3D printing than boxing with language. The best option will be on-fly switchable UI.

Do you consider to use some online services like https://www.transifex.com or https://www.lingotek.com/ ?

I'm ready to help with polish translation.

**Comment by lordofhyphens** [comment]
@SCiunczyk Pull a copy of the i18n branch, make a new .po file and submit that as a PR against that branch.

This will be an ongoing effort as every addition of text to the UI will require additions to the translations.

**Comment by thiagopeixoto16** [comment]
Please, I can't find i18n branch or anything related... can someone help on this?
![image](https://user-images.githubusercontent.com/4227271/54464227-8bcfff00-4754-11e9-83e3-a35cf9748be3.png)
